silkworm/db/snapshots/index/bloom_filter.cpp

```diff
@@ -0,0 +1,206 @@
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+#include "bloom_filter.hpp"
+
+#include <numbers>
+#include <random>
+#include <sstream>
+#include <string_view>
+
+#include <openssl/evp.h>
+
+#include <silkworm/core/common/bytes_to_string.hpp>
+#include <silkworm/core/common/endian.hpp>
+#include <silkworm/core/common/util.hpp>
+
+namespace silkworm::snapshots::index {
+
+using namespace std::numbers;
+
+//! The minimum Bloom filter bits count
+static constexpr size_t kMinimumBitsCount{2};
+
+//! kRotation sets how much to rotate the hash on each filter iteration.
+//! This is somewhat randomly set to a prime on the lower segment of 64.
+constexpr size_t kRotation{17};
+
+constexpr size_t kRotationOf64{64 - kRotation};
+
+//! The magic header used in serialization format for version v2
+static const ByteView kMagicHeader{string_view_to_byte_view("\0\0\0\0\0\0\0\0v02\n"sv)};
+
+uint64_t BloomFilter::optimal_bits_count(uint64_t max_key_count, double p) {
+    return static_cast<uint64_t>(std::ceil(-static_cast<double>(max_key_count) * std::log(p) / (ln2 * ln2)));
+}
+
+std::unique_ptr<BloomFilter> BloomFilter::read_from(std::istream& index_input_stream) {
+    auto filter = std::make_unique<BloomFilter>(kMinimumBitsCount);  // bits count doesn't matter here, it will be overwritten
+    index_input_stream >> *filter;
+    return filter;
+}
+
+BloomFilter::BloomFilter(uint64_t bits_count)
+    : BloomFilter(bits_count, new_random_keys()) {}
+
+BloomFilter::BloomFilter(uint64_t bits_count, KeyArray keys)
+    : bits_count_(bits_count),
+      keys_(keys),
+      bits_((bits_count + 63) / 64, 0) {
+    check_bits_count(bits_count);
+}
+
+BloomFilter::BloomFilter(uint64_t max_key_count, double p)
+    : BloomFilter(optimal_bits_count(max_key_count, p)) {}
+
+void BloomFilter::add_hash(uint64_t hash) {
+    for (size_t n = 0; n < kHardCodedK; ++n) {
+        hash = ((hash << kRotation) | (hash >> kRotationOf64)) ^ keys_[n];
+        const uint64_t i = hash % bits_count_;
+        bits_[i >> 6] |= uint64_t{1} << (i & 0x3F);
+    }
+    ++inserted_count_;
+}
+
+bool BloomFilter::contains_hash(uint64_t hash) {
+    uint64_t r{1};
+    for (size_t n = 0; n < kHardCodedK; ++n) {
+        hash = ((hash << kRotation) | (hash >> kRotationOf64)) ^ keys_[n];
+        const uint64_t i = hash % bits_count_;
+        r &= (bits_[i >> 6] >> (i & 0x3F)) & uint64_t{1};
+    }
+    return r != 0;
+}
+
+void BloomFilter::check_bits_count(uint64_t bits_count) {
+    if (bits_count < kMinimumBitsCount) {
+        throw std::runtime_error{"number of bits must be >= " + std::to_string(kMinimumBitsCount) +
+                                 " (was " + std::to_string(bits_count) + ")"};
+    }
+}
+
+BloomFilter::KeyArray BloomFilter::new_random_keys() {
+    // TODO(canepat) std::random_device is not guaranteed to be a CS-PRNG
```

Comment:
what does it mean? what's the implication?

Reply:
It means that we should use a PRNG relying on sources of randomness provided by the OS to improve the robustness of the filter to cryptographic attacks. I'll add a comment